yarn-project/types/src/l1_to_l2_message.ts

```diff
@@ -17,6 +17,79 @@ export interface L1ToL2MessageSource {
   getPendingL1ToL2Messages(take: number): Promise<L1ToL2Message[]>;
 }
 
+/**
+ * A simple in-memory implementation of an L1 to L2 message store
+ * that handles message duplication.
+ */
+export class L1ToL2MessageStore {
+  /**
+   * A map containing the message key to the corresponding L1 to L2
+   * messages (and the number of times the message has been seen).
+   */
+  private store: Map<Fr, L1ToL2MessageAndCount> = new Map();
+
+  constructor() {}
+
+  addMessage(entryKey: Fr, msg: L1ToL2Message) {
```

Comment:
The entryKey can be compute from message. Would be be more clean to just have `addMEssage(msg: L1ToL2Message)` and compute it in there instead? Not a big deal though.